### PR TITLE
Add feature for formatting only VCS changed text

### PR DESCRIPTION
**Pull request by asarium**
This is similar to what is available in the built-in format action. It
can be enabled using a check box in the settings which is disabled by
default.

### CODE REVIEW

File: src/io/probst/idea/clangformat/ClangFormatAction.java [hunk 1]
```diff
@@ -13,29 +14,24 @@
...
 import javax.xml.bind.annotation.XmlRootElement;
 import javax.xml.bind.annotation.XmlValue;
-import java.io.File;
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.InputStreamReader;
-import java.io.OutputStream;
-import java.io.OutputStreamWriter;
+import javax.xml.soap.Text;
+import java.io.*;
```
Comment (mprobst) [inline src/io/probst/idea/clangformat/ClangFormatAction.java]: do not use * imports

File: src/io/probst/idea/clangformat/ClangFormatAction.java [hunk 2]
```diff
@@ -13,29 +14,24 @@
 import com.intellij.openapi.editor.Editor;
 import com.intellij.openapi.project.DumbAware;
 import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.TextRange;
```
Comment (mprobst) [inline src/io/probst/idea/clangformat/ClangFormatAction.java:16]: do you have a bunch of unused imports here?

File: src/io/probst/idea/clangformat/ClangFormatAutoAction.java [hunk 3]
```diff
@@ -1,26 +1,66 @@
...
+        // IntelliJ reports a cursor at the end of the file as being at file length + 1, which breaks
+        // clang-format.
+        int docLength = document.getTextLength() - 1;
+        int selectionStart = Math.min(editor.getSelectionModel().getSelectionStart(), docLength);
+        int selectionLength = Math.min(editor.getSelectionModel().getSelectionEnd(), docLength) - selectionStart;
+
+        // Formatting the VCS changed text will only be done if nothing is selected. Otherwise you can't force a reformat
+        // of some code that was not changed from the VCS version
+        if (selectionLength <= 0 && settings.updateOnlyChangedText) {
+            // This case only formats the text changed in the VCS. If the diff is too large or we are not in a VCS file
```
Comment (mprobst) [inline src/io/probst/idea/clangformat/ClangFormatAutoAction.java:62]: can you extract a method for this branch here? it's conceptually different enough to warrant its own place and name.

File: src/io/probst/idea/clangformat/ClangFormatAction.java (hunk 1 above)
Comment (asarium) [reply src/io/probst/idea/clangformat/ClangFormatAction.java]: I expanded all `*` important and removed all other unused imports I could find.

File: src/io/probst/idea/clangformat/ClangFormatAutoAction.java (hunk 3 above)
Comment (asarium) [reply src/io/probst/idea/clangformat/ClangFormatAutoAction.java:62]: I extracted the function.

**Review by mprobst — APPROVED** [review]

**Pull request merged**